applications/GeoMechanicsApplication/custom_elements/transient_Pw_line_element.h

```diff
@@ -63,34 +71,72 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) TransientPwLineElement : public Elem
         rResult = Geo::DofUtilities::ExtractEquationIdsFrom(GetDofs());
     }
 
+    void Initialize(const ProcessInfo& rCurrentProcessInfo) override
+    {
+        if (const std::size_t number_integration_points =
+                GetGeometry().IntegrationPointsNumber(GetIntegrationMethod());
+            mRetentionLawVector.size() != number_integration_points) {
+            mRetentionLawVector.resize(number_integration_points);
+        }
+        for (unsigned int i = 0; i < mRetentionLawVector.size(); ++i) {
+            mRetentionLawVector[i] = RetentionLawFactory::Clone(GetProperties());
+            mRetentionLawVector[i]->InitializeMaterial(
+                GetProperties(), GetGeometry(),
+                row(GetGeometry().ShapeFunctionsValues(GetIntegrationMethod()), i));
+        }
+    }
+
+    void InitializeSolutionStep(const ProcessInfo&) override
+    {
+        RetentionLaw::Parameters RetentionParameters(this->GetProperties());
+        for (const auto& retention_law : mRetentionLawVector) {
+            retention_law->InitializeSolutionStep(RetentionParameters);
+        }
+    }
+
+    void FinalizeSolutionStep(const ProcessInfo&) override
+    {
+        RetentionLaw::Parameters RetentionParameters(this->GetProperties());
+        for (const auto& retention_law : mRetentionLawVector) {
+            retention_law->FinalizeSolutionStep(RetentionParameters);
+        }
+    }
+
     void CalculateLocalSystem(MatrixType&        rLeftHandSideMatrix,
                               VectorType&        rRightHandSideVector,
                               const ProcessInfo& rCurrentProcessInfo) override
     {
         KRATOS_TRY
 
-        Vector det_J_container;
-        GetGeometry().DeterminantOfJacobian(det_J_container, this->GetIntegrationMethod());
-        GeometryType::ShapeFunctionsGradientsType dN_dX_container =
-            GetGeometry().ShapeFunctionsLocalGradients(this->GetIntegrationMethod());
-        std::transform(dN_dX_container.begin(), dN_dX_container.end(), det_J_container.begin(),
-                       dN_dX_container.begin(), std::divides<>());
-        const Matrix& r_N_container = GetGeometry().ShapeFunctionsValues(GetIntegrationMethod());
+        rRightHandSideVector = CalculateFluidBodyVector();
+        rLeftHandSideMatrix  = ZeroMatrix{TNumNodes, TNumNodes};
 
-        const auto integration_coefficients = CalculateIntegrationCoefficients(det_J_container);
-        const auto permeability_matrix = CalculatePermeabilityMatrix(dN_dX_container, integration_coefficients);
-        const auto compressibility_matrix =
-            CalculateCompressibilityMatrix(r_N_container, integration_coefficients);
+        for (const auto& rContribution : mContributions) {
+            const auto calculator = CreateCalculator(rContribution, rCurrentProcessInfo);
```

Comment:
What is the benefit to create calculators every time?

Reply:
I guess `Initialize` is called before `CalculateLocalSystem` and it is possible to create the calculators there. But then the memory will be always taken for all elements.